DietPi-Software | Add Portainer

Joulinar opened the pull request:
DietPi-Software | Portainer: Add new software title

**Status**:  Ready
- [x] Installation of Portainer
- [x] Installation of Portainer together with dependency - Docker
- [x] re-installation + upgrade of Portainer
- [x] un-installation of Portainer
- [x] un-installation of Docker + Portainer

**Devices to be tested**:
- [x] RPi1B 32bit Buster
- [x] RPi3B+ 32bit Buster
- [x] VM 64bit Buster
- [x] VM 64bit Stretch
- [x] RPi3B+ 64bit Buster

**Reference**: https://github.com/MichaIng/DietPi/issues/2848

**Commit list/description**:
- DietPi-Software | Add Portainer, initial setup
- DietPi-Software | Portainer: some code fixes
- DietPi-Software | Portainer: shift installation steps into configuration section
- DietPi-Software | Docker: remove configuration file daemon.json during un-installation
- DietPi-Software | Portainer: adjust wording
- DietPi-Software | Portainer: add clear identification of container and image files for portainer-ce
- DietPi-Software | Portainer: Change software ID to avoid conflicts with other new software title
- DietPi-Software | Portainer: fixes and adjustments
- DietPi-Software | Portainer: change port to 9002 as LMS is using port 9000 already
- DietPi-Software | Portainer: disable support for Bullseye, as there is still no Docker repo suite
- DietPi-Software | Portainer: change SOFTWARE_CATEGORY_INDEX to 8
- DietPi-Software | Portainer: change dependency for Docker to avoid issues if kernel was update right before installation

Commit:
DietPi-Software | Add Portainer

 DietPi-Software | Portainer: Add new software title

Joulinar commented:
ok already to late today, will fix the "Quotes" tomorrow 🙄

Joulinar committed:
DietPi-Software | Portainer

DietPi-Software | Portainer: some code fixes

Joulinar commented:
ok found an issue on deploying a container on new fresh installation. Because `/etc/docker/daemon.json` is missing, which is going to be created on a later step within Docker configuration. Need to shift activities from left to right to get correct order and dependencies 😃

Joulinar committed:
v6.34

+ DietPi-Software | Portainer: shift installation steps into configuration section
+ DietPi-Software | Docker: Docker configuration file daemon.json during un-installation

Joulinar commented:
ok whole installation process is shifted into configuration section now, to ensure Docker is installed correctly before deploying a container. As well I like to avoid to many restarts of `docker.service`.

Joulinar committed:
v6.34

+ DietPi-Software | Portainer: adjust wording

Joulinar commented:
Testing done from my side. So fare no issues detected. But I have 2 questions

- do you think it would make sense to add Portainer Agent as well? It's not a big deal 😃 
- I noticed some file system not going to be deleted on Docker un-installation. Not sure if this is intended.
```
/var/lib/containerd/
/opt/containerd/
```

Commit:
v6.34

ietPi-Software | Portainer: add clear identification of container and image files for portainer-ce

Commit:
v6.34

+ DietPi-Software | Portainer: Change software ID to avoid conflicts with other new software title

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12236,7 +12253,33 @@ _EOF_
+			# Restart Docker service to be able to deploy the container
+			G_DIETPI-NOTIFY 2 'Docker will be restarted to be able to deploy the container.'
+			G_EXEC systemctl daemon-reload
+			G_EXEC systemctl restart docker.service
```
The `.service` ending is btw optional, as it is assumed automatically by systemctl.

Joulinar replied:
ok will remove `.service`

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12236,7 +12253,33 @@ _EOF_
+			if [[ -d '/mnt/dietpi_userdata/docker-data/volumes/portainer_data' ]]; then
+				G_DIETPI-NOTIFY 2 'Portainer detected, container will be dropped before reinstallation.'
+				G_EXEC docker rm "$(docker ps -a | grep "portainer/portainer-ce" | awk '{print $1}')" --force
+				G_EXEC docker rmi "$(docker images -a | grep "portainer/portainer-ce" | awk '{print $3}')"
```
`grep` and `awk` can always be merged into a single call. Also, instead of `awk`, we use `mawk` which is the faster Debian default. Only in case it is installed, the slower (but feature-richer) `gawk` will replace the `awk` alternative symlink. So using `mawk` explicitly assures consistency across all systems and faster execution:
```
G_EXEC docker rm "$(docker ps -a | mawk '/portainer\/portainer-ce/{print $1}')" --force
G_EXEC docker rmi "$(docker images -a | mawk '/portainer\/portainer-ce/{print $3}')"
```

Same below in uninstall step

Joulinar replied:
ok will change it that way. Your code is working well on my small test.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12236,7 +12253,33 @@ _EOF_
+
+			# Deploy the portainer container
+			G_DIETPI-NOTIFY 2 'Portainer will be deployed now. This could take a while...'
+			G_EXEC docker run -d -p 9000:9000 --name=portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce
```
I guess this contains some interesting output? In case, it can be shown on console via:
```
G_EXEC_OUTPUT=1 G_EXEC docker run ...
```

Joulinar replied:
ok this will show download and extraction of files. That's how the output will looks like

```
[ INFO ] DietPi-Software | Portainer will be deployed now. This could take a while...
[ INFO ] DietPi-Software | docker run -d -p 9000:9000 --name=portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce, please wait...
Unable to find image 'portainer/portainer-ce:latest' locally
latest: Pulling from portainer/portainer-ce
d1e017099d17: Pulling fs layer
b0718b1ef1b0: Pulling fs layer
d1e017099d17: Verifying Checksum
d1e017099d17: Download complete
d1e017099d17: Pull complete
b0718b1ef1b0: Verifying Checksum
b0718b1ef1b0: Download complete
b0718b1ef1b0: Pull complete
Digest: sha256:0ab9d25e9ac7b663a51afc6853875b2055d8812fcaf677d0013eba32d0bf0e0d
Status: Downloaded newer image for portainer/portainer-ce:latest
b457069bc66d4efc736bef59afdbb64266010d746ca25e27eb99bcba520ac5e6
[  OK  ] DietPi-Software | docker run -d -p 9000:9000 --name=portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce
```

MichaIng replied:
Especially if this takes a bid longer it might be good to know the current step 👍. And Docker users might like to see something familiar or on the other hand become familiar with the output when installing own containers 😉.

Joulinar replied:
execution depends on device. On a RPi3 or 4 it's quite fast but on RPiB, will it take some minutes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -14864,6 +14907,30 @@ _EOF_
 			# Remove default unused folder
 			[[ -d '/var/lib/docker' ]] && rm -R /var/lib/docker
+			# Remove configuration file   
+			[[ -f '/etc/docker/daemon.json' ]] && rm -R /etc/docker/
```
I guess the empty dir (or if anything else is contained) could be removed as well:
```
[[ -d '/etc/docker' ]] && rm -R /etc/docker
```

Joulinar replied:
ok going to check for the directory and remove it if existing

Commit:
v6.34

+ DietPi-Software | Portainer: fixes and adjustments

Joulinar commented:
ok I did the adjustments. We just need to agree on the output during installation. If it's something usful or more confusing 😃

MichaIng commented:
Ah once last thing, Logitech Media Server already listens to port **9000** by default. We can switch to **9001** with `docker run -d -p 9001:9001`, right?
_________
Aside of this, it works perfectly fine, great work!

Joulinar commented:
yup will change it to 9001

MichaIng commented:
The following works:
```
G_EXEC_OUTPUT=1 G_EXEC docker run -d -p 9001:9000 --name=portainer --restart=always -v /run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce
```
Container internal port needs to stay 9000, and this includes mounting the actual `/run/docker.sock` instead of symlink `/var/run/docker.sock`. Not sure if Docker-internal `/run` as well works?

nonnorm commented:
As per #3951, make sure to change `aSOFTWARE_CATEGORY_INDEX` to 9.

Joulinar commented:
Ok will do changes later on. Family time right now 😄

Joulinar commented:
> As per #3951, make sure to change aSOFTWARE_CATEGORY_INDEX to 9.

I guess it would need to be 8 instead of 9, right?
`aSOFTWARE_CATEGORY_INDEX[$software_id]=8`

MichaIng commented:
Yes exactly, 9 is it now, 8 shall it be 😄.

nonnorm commented:
My bad. Yes, 8.

Joulinar commented:
ok found one issue. Thx to the kernel update. We have a check that prevent docker from being installed if there was a kernel update right before. YEAHHHHH. On DietPi-Software, if you select Portainer only (and Docker is not installed already), the Docker-Kernel check will be baypassed because Docker will be marked as dependence later one

MichaIng commented:
Good find. So that check needs to simply set both to 0.

I see that being an issue on reinstalls, actually: For reinstalls, install flag is set to 1 as well and the check sets it to 0 if the kernel got just upgraded which will be stored like that at the end of the install run. So that check would need to be done as well during reinstall input handling, but that is very ugly. I'd leave it with that flaw for now. Final reasonable solution could be to only store the state of actually just installed software titles to the file, and skip the ones with 0. On uninstall the ones with state -1 would then simply removed from the file. This lowers as well the file size and increases read/write speeds. Only not anymore with this release, as it requires longer testing, I think.

Joulinar commented:
I skip Docker on Portainer reinstall, means container will be deployed only

Problem, the Docker-Kernel check is not even touched if Portainer is selected only. Because Docker will be flagged to be installed way behind that check. I guess I will remove `SOFTWARE_REQUIRES_DOCKER` and move Docker dependency for Portainer before that check.

Joulinar commented:
one more, I see you disabled Docker on Bullseye, right?

```
		# - Bullseye
		aSOFTWARE_AVAIL_G_DISTRO[$software_id,6]=0
```

should we do same on Portainer?

MichaIng commented:
Let me see if this is still required.

Joulinar commented:
Ahh we should not use 9001 as port because this is used for Portainer agent. We would need an alternative port

MichaIng commented:
Then `9002`?

About Bullseye, there is still no Docker repo suite for it: https://download.docker.com/linux/debian/dists/
So Portainer should then be disabled as well as it would pull in Docker after it was unmarked by distro check, AFAIK.

Joulinar commented:
ok will change to 9002 and disable docker on Bullseye

MichaIng commented:
Actually we can do much better what the Docker installer does: Install key + install repo + install docker-ce package, that's it. The installer suffered from issues in the past, e.g. wrong/missing Buster support and adds at least the key in a way that should not be done anymore. Time to create an own APT repo function and add those in a consistent modern way... in next development cycle 😄.

Commit:
v6.34

+ DietPi-Software | Portainer: change port to 9002 as LMS is using port 9000 already
+ DietPi-Software | Portainer: disable support for Bullseye, as there is still no Docker repo suite
+ DietPi-Software | Portainer: change SOFTWARE_CATEGORY_INDEX to 8
+ DietPi-Software | Portainer: change dependency for Docker to avoid issues if kernel was update right before installation

Joulinar commented:
ok this should be working.

@MichaIng pls  can you habe a look again

MichaIng committed:
Merge branch 'dev' into portainer

MichaIng committed:
v6.34

+ DietPi-Software | Fix alignment, trailing spaces and doubled empty lines

MichaIng commented:
Okay, while resolving the conflict I recognised a lot of alignment issues, tailing spaces and doubled empty lines (not part of your PR) and fixed that as well 😉.

Jep the Docker install prevention on Bullseye is now reliable. It's a bid bad that it is first marked and and info printed, and then immediately unmarked with the kernel mismatch prompt. However, that's the best we can do for now without larger restructure.

Joulinar commented:
yeah quite some shifting on the code due to merge from `dev`. To many software developments for v6.34 🤣

MichaIng commented:
Okay, another install and uninstall test and works all fine now. I'll merge it.
